File: src/rac/services/export.py
"""Corpus projections — `rac export` and its `--documents` / `--graph` / `--okf` modes.

Three deterministic, read-only projections of a corpus, each composed from the
same Core services (traversal, identity/aliases, relationship resolution) so a
single walk yields a stable public payload (ADR-007):

- :func:`build_corpus_export` — the viewer/Portal payload (:class:`CorpusExport`):
  every classified artifact with its rendered HTML body and flattened, untyped
  ``relates-to`` edges.
- :func:`build_documents_export` — the ingestion projection (:class:`DocumentsExport`):
  one Markdown-body document per artifact for memory/RAG backends.
- :func:`build_graph_export` — the typed graph (:class:`GraphExport`, ADR-074):
  nodes plus edges carrying their registry edge kind and direction.

Determinism (ADR-002): no timestamps, no environment-dependent fields (bar the
producing CLI's version, captured into the model at build time), artifacts and
nodes in sorted-path order, edges sorted by a total key — two exports of the
same tree are byte-identical.

Shared gate: ``spec_for(type) is None`` (an unknown file) is excluded from the
projection, but the file still feeds reference resolution — an edge resolves
here exactly when ``relationships --validate`` reports it resolved.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import PurePath
from typing import NamedTuple

from markdown_it import MarkdownIt

import rac
from rac.core.artifacts import ArtifactSpec, spec_for
from rac.core.corpus import CorpusEntry, walk_corpus
from rac.core.frontmatter import split_frontmatter
from rac.core.identity import artifact_identifier, artifact_identifiers
from rac.core.models import Product
from rac.core.relationship_types import edge_spec

from .init import load_ticketing_provider
from .inspect import canonical_value
from .relationships import relationships_from_corpus

# The single edge type the viewer payload emits; richer typing lives in the
# graph projection, not here.
EDGE_TYPE = "relates-to"

# Exported status for an artifact whose ``## Status`` section is missing or empty
# (the viewer contract requires a string where ``rac inspect`` would omit it).
STATUS_ABSENT = "unknown"


@dataclass
class ExportArtifact:
    """One artifact in the viewer payload (viewer contract v1)."""

    id: str
    aliases: list[str]
    type: str
    status: str
    title: str
    path: str
    body_html: str
    # OKF-reserved descriptive labels (ADR-050), carried for the OKF bundle
    # projection only. Deliberately absent from ``to_dict``: the frozen JSON
    # contract (ADR-007) predates tags and stays unchanged until a versioned add.
    tags: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "aliases": self.aliases,
            "type": self.type,
            "status": self.status,
            "title": self.title,
            "path": self.path,
            "body_html": self.body_html,
        }


@dataclass
class ExportRelationship:
    """One flattened ``relates-to`` edge, read "``from`` ``type`` ``to``".

    ``from_`` serializes as ``"from"`` (the keyword cannot name a field). ``to``
    is a canonical identifier when the reference resolves, else the literal
    reference text preserved verbatim — the viewer renders unresolved targets
    rather than dropping them.
    """

    from_: str
    to: str
    type: str = EDGE_TYPE

    def to_dict(self) -> dict:
        return {"from": self.from_, "to": self.to, "type": self.type}


@dataclass
class CorpusExport:
    """Deterministic viewer/Portal payload.

    ``to_dict`` is the stable JSON contract (ADR-007); ``schema_version`` is the
    string ``"1"``. ``rac_version`` is captured from the producing CLI at build
    time so the payload value itself carries no live environment lookup.
    """

    corpus_name: str
    rac_version: str
    artifacts: list[ExportArtifact] = field(default_factory=list)
    relationships: list[ExportRelationship] = field(default_factory=list)

    @property
    def artifact_count(self) -> int:
        return len(self.artifacts)

    def to_dict(self) -> dict:
        return {
            "schema_version": "1",
            "corpus": {
                "name": self.corpus_name,
                "rac_version": self.rac_version,
                "artifact_count": self.artifact_count,
            },
            "artifacts": [artifact.to_dict() for artifact in self.artifacts],
            "relationships": [edge.to_dict() for edge in self.relationships],
        }


@dataclass
class ExportDocument:
    """One artifact as an ingestion-ready document.

    ``text`` is the artifact's Markdown *body* (frontmatter stripped), not
    rendered HTML: memory/RAG backends embed text, and markup would be noise.
    The artifact is the atomic unit (ADR-004, ADR-010) — one document each,
    never chunked.
    """

    id: str
    type: str
    status: str
    title: str
    text: str
    aliases: list[str]
    path: str
    tags: list[str] = field(default_factory=list)

    def to_dict(self, source: str) -> dict:
        """One JSONL record; ``source`` namespaces the corpus (e.g. a container tag)."""
        return {
            "schema_version": "1",
            "id": self.id,
            "type": self.type,
            "status": self.status,
            "title": self.title,
            "text": self.text,
            "metadata": {
                "path": self.path,
                "aliases": self.aliases,
                "tags": self.tags,
                "source": source,
            },
        }


@dataclass
class DocumentsExport:
    """Deterministic ingestion projection, serialized as JSON Lines.

    One record per classified artifact, sorted-path order, no timestamps
    (ADR-002). Additive and separate from the viewer JSON (ADR-007). Each record
    carries the canonical ``id`` (re-fetch hook) and ``status`` (so a retired
    artifact is filterable on read).
    """

    corpus_name: str
    documents: list[ExportDocument] = field(default_factory=list)

    @property
    def document_count(self) -> int:
        return len(self.documents)

    def to_records(self) -> list[dict]:
        return [doc.to_dict(self.corpus_name) for doc in self.documents]


@dataclass
class GraphNode:
    """One artifact as a graph node."""

    id: str
    type: str
    status: str
    title: str

    def to_dict(self) -> dict:
        return {"id": self.id, "type": self.type, "status": self.status, "title": self.title}


@dataclass
class GraphEdge:
    """One typed relationship edge (ADR-074).

    ``type`` is the registry edge kind and ``directed`` follows the registry
    (``supersedes`` is directed; the ``related_*`` edges are not). ``resolved``
    is False when the reference does not resolve, in which case ``target`` is the
    literal reference text — no phantom node is invented. ``external`` is True for
    an external-reference edge (``related_tickets``, ADR-087; ``verified_by``,
    ADR-096) whose target is not an in-corpus artifact, distinguishing it from a
    dangling in-corpus link. ``provider`` carries the configured ticketing system
    (ADR-088) only for provider-backed external edges — ``verified_by`` is
    external and directed but not provider-tagged.
    """

    source: str
    target: str
    type: str
    directed: bool
    resolved: bool
    external: bool = False
    provider: str | None = None

    def to_dict(self) -> dict:
        return {
            "source": self.source,
            "target": self.target,
            "type": self.type,
            "directed": self.directed,
            "resolved": self.resolved,
            "external": self.external,
            "provider": self.provider,
        }


@dataclass
class GraphExport:
    """Deterministic typed node+edge projection (ADR-074).

    Surfaces the *typed* relationship graph for graph backends, unlike the
    viewer JSON's flattened ``relates-to`` edges (unchanged). Nodes in
    sorted-path order, edges sorted by ``(source, type, target)``, no timestamps.
    """

    corpus_name: str
    nodes: list[GraphNode] = field(default_factory=list)
    edges: list[GraphEdge] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "schema_version": "1",
            "source": self.corpus_name,
            "nodes": [node.to_dict() for node in self.nodes],
            "edges": [edge.to_dict() for edge in self.edges],
        }


class _Projection(NamedTuple):
    """A classified entry ready to project: path, entry, its spec, canonical id."""

    path: str
    entry: CorpusEntry
    spec: ArtifactSpec
    canonical: str


def _project(entries: list[CorpusEntry]) -> tuple[dict[str, str], list[_Projection]]:
    """Split a walked corpus into the resolution index and the projectable artifacts.

    The returned ``canonical_by_path`` maps *every* entry's path to its canonical
    id — unknown files included, because a reference that resolves to an unknown
    file must still name it in an edge. The list holds only classified entries
    (``spec is not None``), in the walk's sorted-path order, so all three builders
    share one gate and one loop instead of three near-duplicates.
    """
    canonical_by_path: dict[str, str] = {}
    projections: list[_Projection] = []
    for entry in entries:
        path = str(entry.path)
        spec = spec_for(entry.artifact_type)  # None for unknown
        canonical = artifact_identifier(entry.product, spec, path)
        canonical_by_path[path] = canonical
        if spec is not None:
            projections.append(_Projection(path, entry, spec, canonical))
    return canonical_by_path, projections


def _corpus_name(directory: str) -> str:
    """The directory's basename, stable relative to the argument as given.

    Trailing separators are stripped so ``rac/`` and ``rac`` name one corpus; the
    path is not resolved against the filesystem, so output is independent of the
    working directory.
    """
    return PurePath(directory.rstrip("/")).name or directory


def _status(product: Product, spec: ArtifactSpec) -> str:
    """The lifecycle status in inspect's canonical spelling, else ``STATUS_ABSENT``.

    Canonicalizes the ``## Status`` value against the type's declared metadata
    (``accepted`` -> ``Accepted``); the rendered body still keeps the literal
    source spelling.
    """
    body = product.sections.get("status")
    if not body:
        return STATUS_ABSENT
    return canonical_value(body, spec.metadata.get("status", ())) or STATUS_ABSENT


def _body_markdown(path: str) -> str:
    """The Markdown body after the frontmatter envelope (a fresh read).

    The parsed ``Product`` keeps sections, not the raw body, and a ``CorpusEntry``
    does not carry the source text — so the file is re-read here. This is the one
    place export reaches back to disk, shared by the HTML and document bodies.
    """
    with open(path, encoding="utf-8") as fh:
        return split_frontmatter(fh.read()).body


def _render_body(path: str, md: MarkdownIt) -> str:
    """Render the artifact's Markdown body to HTML (raw HTML escaped, never run)."""
    return md.render(_body_markdown(path))


def build_corpus_export(directory: str, recursive: bool = True) -> CorpusExport:
    """Export every classified artifact under ``directory`` as the viewer payload."""
    entries = list(walk_corpus(directory, recursive=recursive))
    # The commonmark preset enables raw HTML (the spec includes it); the Portal
    # trust model requires it off, so source HTML arrives escaped, not executed
    # (ADR-059: one parser instance per build).
    md = MarkdownIt("commonmark", {"html": False})
    canonical_by_path, projections = _project(entries)

    artifacts: list[ExportArtifact] = []
    for path, entry, spec, canonical in projections:
        meta = entry.product.metadata
        artifacts.append(
            ExportArtifact(
                id=canonical,
                aliases=artifact_identifiers(entry.product, spec, path),
                type=entry.artifact_type,
                status=_status(entry.product, spec),
                title=entry.product.title or canonical,
                path=path,
                body_html=_render_body(path, md),
                tags=meta.tags if meta else [],
            )
        )

    edges = [
        ExportRelationship(
            from_=canonical_by_path[rel.source_path],
            to=(canonical_by_path[rel.resolved_path] if rel.resolved_path else rel.target),
        )
        for rel in relationships_from_corpus(entries)
    ]
    edges.sort(key=lambda edge: (edge.from_, edge.to))

    return CorpusExport(
        corpus_name=_corpus_name(directory),
        rac_version=rac.__version__,
        artifacts=artifacts,
        relationships=edges,
    )


def build_documents_export(directory: str, recursive: bool = True) -> DocumentsExport:
    """Project every classified artifact under ``directory`` as an ingestion document.

    Same gate as :func:`build_corpus_export`, but emits the Markdown body and the
    verify-in-Lore metadata rather than the viewer's HTML.
    """
    entries = list(walk_corpus(directory, recursive=recursive))
    _, projections = _project(entries)

    documents = [
        ExportDocument(
            id=canonical,
            type=entry.artifact_type,
            status=_status(entry.product, spec),
            title=entry.product.title or canonical,
            text=_body_markdown(path),
            aliases=artifact_identifiers(entry.product, spec, path),
            path=path,
            tags=(entry.product.metadata.tags if entry.product.metadata else []),
        )
        for path, entry, spec, canonical in projections
    ]
    return DocumentsExport(corpus_name=_corpus_name(directory), documents=documents)


def build_graph_export(directory: str, recursive: bool = True) -> GraphExport:
    """Project the corpus as typed nodes and edges (ADR-074).

    Nodes are the classified artifacts; edges carry the registry edge kind and
    its direction. Resolved targets become canonical-id edges; unresolved
    references keep their literal target with ``resolved: False`` rather than
    being dropped.
    """
    entries = list(walk_corpus(directory, recursive=recursive))
    # The configured ticketing provider tags provider-backed external edges
    # (ADR-088); read once for the whole build.
    provider = load_ticketing_provider(directory)
    canonical_by_path, projections = _project(entries)

    nodes = [
        GraphNode(
            id=canonical,
            type=entry.artifact_type,
            status=_status(entry.product, spec),
            title=entry.product.title or canonical,
        )
        for _path, entry, spec, canonical in projections
    ]

    edges: list[GraphEdge] = []
    for rel in relationships_from_corpus(entries):
        kind = edge_spec(rel.relationship)
        target = (
            canonical_by_path[rel.resolved_path] if rel.resolved_path is not None else rel.target
        )
        edges.append(
            GraphEdge(
                source=canonical_by_path[rel.source_path],
                target=target,
                type=rel.relationship,
                directed=kind.directional if kind else False,
                resolved=rel.resolved_path is not None,
                external=kind.external if kind else False,
                provider=provider if (kind and kind.external_provider) else None,
            )
        )
    edges.sort(key=lambda edge: (edge.source, edge.type, edge.target))

    return GraphExport(corpus_name=_corpus_name(directory), nodes=nodes, edges=edges)

File: rust/rac-engine/src/export.rs
//! Corpus export (`rac.services.export`) — deterministic viewer/graph/documents
//! projections of a corpus. One walk, shared across projections; no timestamps.

use crate::identity::{artifact_identifier, artifact_identifiers};
use crate::markdown::split_frontmatter;
use crate::parse::Artifact;
use crate::pycompat::py_strip;
use crate::relationships::{
    corpus_items, edge_spec, relationships_from_corpus, CorpusItem,
};
use crate::spec::ArtifactSpec;
use crate::validate::load_ticketing_provider;

pub const EDGE_TYPE: &str = "relates-to";
pub const STATUS_ABSENT: &str = "unknown";

/// `_corpus_name(directory)`.
fn corpus_name(directory: &str) -> String {
    let trimmed = directory.trim_end_matches('/');
    let name = trimmed.rsplit('/').next().unwrap_or("");
    if name.is_empty() || name == "." || name == ".." {
        directory.to_string()
    } else {
        name.to_string()
    }
}

fn first_line(raw: &str) -> String {
    for line in raw.split('\n') {
        let s = py_strip(line);
        if !s.is_empty() {
            return s.to_string();
        }
    }
    String::new()
}

/// `canonical_value(raw, allowed)`, on this module's `first_line`.
fn canonical_value(raw: &str, allowed: &[String]) -> String {
    crate::spec::canonical_value(&first_line(raw), allowed)
}

/// `_status(product, spec)`.
fn status(artifact: &Artifact, spec: &ArtifactSpec) -> String {
    let body = match artifact.section("status") {
        Some(b) if !b.is_empty() => b,
        _ => return STATUS_ABSENT.to_string(),
    };
    let allowed: &[String] = spec
        .metadata
        .iter()
        .find(|(k, _)| k == "status")
        .map(|(_, v)| v.as_slice())
        .unwrap_or(&[]);
    let value = canonical_value(body, allowed);
    if value.is_empty() {
        STATUS_ABSENT.to_string()
    } else {
        value
    }
}

/// The Markdown body after the frontmatter envelope, re-read from disk.
///
/// The oracle re-reads in TEXT mode (`open(path, encoding="utf-8")`), which
/// applies universal newlines — `\r\n` and lone `\r` become `\n` — before
/// `split_frontmatter`. Mirror that here.
///
/// The oracle's text-mode read is also STRICT utf-8: a file with invalid
/// bytes CRASHES the oracle uncaught (`UnicodeDecodeError`) even though the
/// classification walk decoded it with `errors="replace"`. Per PORT-CONTRACT
/// decision 3 this port never crashes; export has no per-artifact issue
/// channel, so the divergence-by-design here is "the Rust export simply
/// succeeds" (catalogued in rust/fuzz/pinned/oracle-crashes/).
fn body_markdown(path: &str) -> String {
    let text = crate::pycompat::read_text_universal(path).unwrap_or_default();
    split_frontmatter(&text).body
}

fn tags_of(artifact: &Artifact) -> Vec<String> {
    artifact
        .metadata
        .as_ref()
        .map(|m| m.tags.clone())
        .unwrap_or_default()
}

fn canonical_by_path(items: &[CorpusItem]) -> std::collections::HashMap<String, String> {
    items
        .iter()
        .map(|it| {
            (
                it.path.clone(),
                artifact_identifier(&it.artifact, it.spec, &it.path),
            )
        })
        .collect()
}

// --- viewer JSON -------------------------------------------------------------

pub struct ExportArtifact {
    pub id: String,
    pub aliases: Vec<String>,
    pub artifact_type: String,
    pub status: String,
    pub title: String,
    pub path: String,
    pub body_html: String,
    /// OKF-reserved descriptive labels (ADR-050): carried for the OKF
    /// bundle projection, deliberately NOT in the viewer JSON (ADR-007).
    pub tags: Vec<String>,
}

pub struct ExportRelationship {
    pub from: String,
    pub to: String,
    pub edge_type: String,
}

pub struct CorpusExport {
    pub corpus_name: String,
    pub rac_version: String,
    pub artifacts: Vec<ExportArtifact>,
    pub relationships: Vec<ExportRelationship>,
}

impl CorpusExport {
    pub fn artifact_count(&self) -> usize {
        self.artifacts.len()
    }
}

pub fn build_corpus_export(directory: &str, rac_version: String) -> CorpusExport {
    let items = corpus_items(directory, true);
    let canonical = canonical_by_path(&items);

    let mut artifacts: Vec<ExportArtifact> = Vec::new();
    for it in &items {
        let Some(spec) = it.spec else { continue };
        let canon = canonical[&it.path].clone();
        let title = match &it.artifact.product.title {
            Some(t) if !t.is_empty() => t.clone(),
            _ => canon.clone(),
        };
        artifacts.push(ExportArtifact {
            id: canon,
            aliases: artifact_identifiers(&it.artifact, it.spec, &it.path),
            artifact_type: spec.name.clone(),
            status: status(&it.artifact, spec),
            title,
            path: it.path.clone(),
            body_html: crate::mdhtml::render(&body_markdown(&it.path)),
            tags: tags_of(&it.artifact),
        });
    }

    let mut edges: Vec<ExportRelationship> = relationships_from_corpus(&items)
        .into_iter()
        .map(|rel| {
            let to = match &rel.resolved_path {
                Some(p) => canonical[p].clone(),
                None => rel.target.clone(),
            };
            ExportRelationship {
                from: canonical[&rel.source_path].clone(),
                to,
                edge_type: EDGE_TYPE.to_string(),
            }
        })
        .collect();
    edges.sort_by(|a, b| a.from.cmp(&b.from).then(a.to.cmp(&b.to)));

    CorpusExport {
        corpus_name: corpus_name(directory),
        rac_version,
        artifacts,
        relationships: edges,
    }
}

// --- documents JSONL ---------------------------------------------------------

pub struct ExportDocument {
    pub id: String,
    pub artifact_type: String,
    pub status: String,
    pub title: String,
    pub text: String,
    pub aliases: Vec<String>,
    pub path: String,
    pub tags: Vec<String>,
}

pub struct DocumentsExport {
    pub corpus_name: String,
    pub documents: Vec<ExportDocument>,
}

pub fn build_documents_export(directory: &str) -> DocumentsExport {
    let items = corpus_items(directory, true);
    let mut documents: Vec<ExportDocument> = Vec::new();
    for it in &items {
        let Some(spec) = it.spec else { continue };
        let canon = artifact_identifier(&it.artifact, it.spec, &it.path);
        let title = match &it.artifact.product.title {
            Some(t) if !t.is_empty() => t.clone(),
            _ => canon.clone(),
        };
        documents.push(ExportDocument {
            id: canon,
            artifact_type: spec.name.clone(),
            status: status(&it.artifact, spec),
            title,
            text: body_markdown(&it.path),
            aliases: artifact_identifiers(&it.artifact, it.spec, &it.path),
            path: it.path.clone(),
            tags: tags_of(&it.artifact),
        });
    }
    DocumentsExport {
        corpus_name: corpus_name(directory),
        documents,
    }
}

// --- graph JSON --------------------------------------------------------------

pub struct GraphNode {
    pub id: String,
    pub artifact_type: String,
    pub status: String,
    pub title: String,
}

pub struct GraphEdge {
    pub source: String,
    pub target: String,
    pub edge_type: String,
    pub directed: bool,
    pub resolved: bool,
    pub external: bool,
    pub provider: Option<String>,
}

pub struct GraphExport {
    pub corpus_name: String,
    pub nodes: Vec<GraphNode>,
    pub edges: Vec<GraphEdge>,
}

pub fn build_graph_export(directory: &str) -> GraphExport {
    let items = corpus_items(directory, true);
    let provider = load_ticketing_provider(directory);
    let canonical = canonical_by_path(&items);

    let mut nodes: Vec<GraphNode> = Vec::new();
    for it in &items {
        let Some(spec) = it.spec else { continue };
        let canon = canonical[&it.path].clone();
        let title = match &it.artifact.product.title {
            Some(t) if !t.is_empty() => t.clone(),
            _ => canon.clone(),
        };
        nodes.push(GraphNode {
            id: canon,
            artifact_type: spec.name.clone(),
            status: status(&it.artifact, spec),
            title,
        });
    }

    let mut edges: Vec<GraphEdge> = Vec::new();
    for rel in relationships_from_corpus(&items) {
        let kind = edge_spec(&rel.relationship);
        let external = kind.map(|k| k.external).unwrap_or(false);
        let target = match &rel.resolved_path {
            Some(p) => canonical[p].clone(),
            None => rel.target.clone(),
        };
        let provider_tag = match kind {
            Some(k) if k.external_provider => provider.clone(),
            _ => None,
        };
        edges.push(GraphEdge {
            source: canonical[&rel.source_path].clone(),
            target,
            edge_type: rel.relationship.clone(),
            directed: kind.map(|k| k.directional).unwrap_or(false),
            resolved: rel.resolved_path.is_some(),
            external,
            provider: provider_tag,
        });
    }
    edges.sort_by(|a, b| {
        a.source
            .cmp(&b.source)
            .then(a.edge_type.cmp(&b.edge_type))
            .then(a.target.cmp(&b.target))
    });

    GraphExport {
        corpus_name: corpus_name(directory),
        nodes,
        edges,
    }
}
